File: javascript/sync_agent.js
#!/usr/bin/env node
// sync_agent.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/clawhub-git-sync-agent/scripts/sync_agent.py
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

/**
 * Permanenter ClawHub ↔ Git Sync Agent
 * Multi-Node fähig, stündliche Ausführung
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const crypto = require('crypto');

// Import sync functions
const {
  syncToGit,
  syncToClawhub,
  log,
  validateSkill,
  getFileHash
} = require('./sync_clawhub_git.js');

const CLAWHUB_DIR = "/home/openclaw/.openclaw/workspace/skills";
const GIT_DIR = "/home/openclaw/.openclaw/workspace/git/skills";
const STATE_FILE = "/home/openclaw/.openclaw/workspace/db/sync_state.json";

// Root directory for backups
const BACKUP_ROOT = "/home/openclaw/.openclaw/workspace/backups/sync_agent";

function loadState() {
  /**Lädt den Sync-State*/
  if (fs.existsSync(STATE_FILE)) {
    const data = fs.readFileSync(STATE_FILE, 'utf8');
    return JSON.parse(data);
  }
  return { sync_history: [], pending: [] };
}

function saveState(state) {
  /**Speichert den Sync-State*/
  const dir = path.dirname(STATE_FILE);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
  fs.writeFileSync(STATE_FILE, JSON.stringify(state, null, 2));
}

function getAllSkills() {
  /**Findet nur valide Skill-Verzeichnisse in beiden Verzeichnissen.*/
  const clawhubSkills = new Set();
  const gitSkills = new Set();

  if (fs.existsSync(CLAWHUB_DIR)) {
    const dirs = fs.readdirSync(CLAWHUB_DIR);
    for (const item of dirs) {
      const fullPath = path.join(CLAWHUB_DIR, item);
      const stat = fs.statSync(fullPath);
      if (stat.isDirectory() && !item.startsWith('.') && !item.startsWith('_')) {
        if (fs.existsSync(path.join(fullPath, "SKILL.md"))) {
          clawhubSkills.add(item);
        }
      }
    }
  }

  if (fs.existsSync(GIT_DIR)) {
    const dirs = fs.readdirSync(GIT_DIR);
    for (const item of dirs) {
      const fullPath = path.join(GIT_DIR, item);
      const stat = fs.statSync(fullPath);
      if (stat.isDirectory() && !item.startsWith('.') && !item.startsWith('_')) {
        if (fs.existsSync(path.join(fullPath, "SKILL.md"))) {
          gitSkills.add(item);
        }
      }
    }
  }

  const allSkills = new Set([...clawhubSkills, ...gitSkills]);
  return Array.from(allSkills);
}

function initGitRepo(skillPath, skillName) {
  /**Initialisiert Git-Repo wenn nötig*/
  const gitDir = path.join(skillPath, ".git");
  if (!fs.existsSync(gitDir)) {
    process.chdir(skillPath);
    execSync("git init", { stdio: 'ignore' });
    execSync("git add .", { stdio: 'ignore' });
    execSync(`git commit -m "Initial commit: ${skillName} skill"`, { stdio: 'ignore' });
    log(`Git initialized for ${skillName}`);
  }
}

function backupSkillDir(skillPath, skillName) {
  /**Creates a timestamped tar.gz backup of a skill directory.*/
  if (!fs.existsSync(skillPath)) {
    return;
  }
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').replace('T', '').slice(0, -1);
  const backupDir = path.join(BACKUP_ROOT, timestamp);
  if (!fs.existsSync(backupDir)) {
    fs.mkdirSync(backupDir, { recursive: true });
  }
  const archiveName = `${skillName}_${timestamp}`;
  const archivePath = path.join(backupDir, archiveName);
  
  // Create tar.gz using system tar command
  execSync(`tar -czf "${archivePath}.tar.gz" -C "${skillPath}" .`, { stdio: 'ignore' });
  log(`Backup created for ${skillName} at ${archivePath}.tar.gz`);
}

async function syncSkillBidirectional(skillName, dryRun = false) {
  /**Bidirektionale Synchronisation eines Skills*/
  const clawhubPath = path.join(CLAWHUB_DIR, skillName);
  const gitPath = path.join(GIT_DIR, skillName);

  // Fall 1: Nur in ClawHub → zu Git
  if (fs.existsSync(clawhubPath) && !fs.existsSync(gitPath)) {
    log(`NEW in ClawHub: ${skillName} → syncing to Git`);
    if (!dryRun) {
      backupSkillDir(clawhubPath, `${skillName}_clawhub`);
    }
    if (await syncToGit(skillName, dryRun)) {
      if (!dryRun) {
        initGitRepo(gitPath, skillName);
      }
      return "synced_to_git";
    }

  // Fall 2: Nur in Git → zu ClawHub
  } else if (fs.existsSync(gitPath) && !fs.existsSync(clawhubPath)) {
    log(`NEW in Git: ${skillName} → syncing to ClawHub`);
    if (!dryRun) {
      backupSkillDir(gitPath, `${skillName}_git`);
    }
    if (await syncToClawhub(skillName, dryRun)) {
      return "synced_to_clawhub";
    }

  // Fall 3: In beiden vorhanden → Vergleiche Timestamps
  } else if (fs.existsSync(clawhubPath) && fs.existsSync(gitPath)) {
    // --- MODIFIZIERTE LOGIK: Robusterer Datei-Hash-Vergleich ---

    // Stelle sicher, dass beide als gültige Skills validiert werden
    if (!validateSkill(clawhubPath)) {
      log(`Validation failed for ClawHub skill: ${skillName}`, "ERROR");
      return "error";
    }
    if (!validateSkill(gitPath)) {
      log(`Validation failed for Git skill: ${skillName}`, "ERROR");
      return "error";
    }

    // Berechne Hashes für clawhub und git
    const clawhubHashes = getHashes(clawhubPath);
    const gitHashes = getHashes(gitPath);

    if (JSON.stringify(clawhubHashes) !== JSON.stringify(gitHashes)) {
      log(`Content difference detected for: ${skillName}`);

      // Einfache (aber oft ausreichende) Logik: Wenn clawhub neuer ist, lade hoch.
      // Eine detailliertere Strategie (z.B. welche Version von Git übernehmen)
      // könnte hier implementiert werden, falls nötig.
      // Für jetzt: Wenn sie sich unterscheiden, priorisieren wir ClawHub > Git
      // und aktualisieren Git.

      const clawhubStat = fs.statSync(clawhubPath);
      const gitStat = fs.statSync(gitPath);
      const direction = clawhubStat.mtimeMs >= gitStat.mtimeMs ? "to-git" : "to-clawhub";
      log(`UPDATE: ${skillName} → syncing ${direction}`);
      if (!dryRun) {
        backupSkillDir(clawhubPath, `${skillName}_clawhub`);
        backupSkillDir(gitPath, `${skillName}_git`);
      }
      const sync = direction === "to-git" ? syncToGit : syncToClawhub;
      if (await sync(skillName, dryRun)) {
        if (!dryRun && direction === "to-git") {
          process.chdir(gitPath);
          execSync("git add .", { stdio: 'ignore' });
          execSync(`git commit -m "Sync from ClawHub content diff: ${new Date().toISOString().slice(0, 16).replace('T', ' ')}"`, { stdio: 'ignore' });
        }
        return direction === "to-git" ? "updated_git" : "updated_clawhub";
      } else {
        log(`Failed to sync ${skillName} to Git after content diff`, "ERROR");
        return "error";
      }
    } else {
      log(`Content is identical for: ${skillName}`);
      return "no_change";
    }
  }

  return "no_change";
}

// --- Hinzufügen dieser Hilfsfunktion ---
function getHashes(skillDir) {
  /**Erzeugt ein Dictionary von Datei-Hashes für einen Skill-Ordner.*/
  const hashes = {};

  function walkDir(currentDir) {
    const items = fs.readdirSync(currentDir);
    for (const item of items) {
      const fullPath = path.join(currentDir, item);
      const relativePath = path.relative(skillDir, fullPath);
      
      // Ignoriere .git Verzeichnisse
      if (relativePath.includes('.git')) {
        continue;
      }
      
      const stat = fs.statSync(fullPath);
      if (stat.isDirectory()) {
        walkDir(fullPath);
      } else {
        hashes[relativePath] = getFileHash(fullPath);
      }
    }
  }

  walkDir(skillDir);
  return hashes;
}

async function main() {
  /**Hauptfunktion des Sync-Agents*/
  const args = process.argv.slice(2);
  const dryRun = args.includes('--dry-run') || args.includes('-d');
  
  log("=== ClawHub ↔ Git Sync Agent gestartet ===");

  const state = loadState();
  const allSkills = getAllSkills();
  log(`Gefundene Skills: ${allSkills.length}`);

  const results = {
    synced_to_git: [],
    synced_to_clawhub: [],
    updated_git: [],
    updated_clawhub: [],
    no_change: [],
    errors: []
  };

  for (const skill of allSkills.sort()) {
    try {
      const result = await syncSkillBidirectional(skill, dryRun);
      results[result].push(skill);
    } catch (e) {
      log(`ERROR syncing ${skill}: ${e.message}`, "ERROR");
      results.errors.push(skill);
    }
  }

  // Zusammenfassung
  log("\n=== SYNC ZUSAMMENFASSUNG ===");
  log(`Neu in Git: ${results.synced_to_git.length} - ${results.synced_to_git}`);
  log(`Neu in ClawHub: ${results.synced_to_clawhub.length} - ${results.synced_to_clawhub}`);
  log(`Git aktualisiert: ${results.updated_git.length} - ${results.updated_git}`);
  log(`ClawHub aktualisiert: ${results.updated_clawhub.length} - ${results.updated_clawhub}`);
  log(`Keine Änderung: ${results.no_change.length}`);
  log(`Fehler: ${results.errors.length} - ${results.errors}`);

  // Ein Dry-Run bleibt vollständig nicht-mutierend (abgesehen vom Audit-Log).
  if (!dryRun) {
    if (!state.sync_history) {
      state.sync_history = [];
    }
    state.sync_history.push({
      timestamp: new Date().toISOString(),
      results: results
    });
    // Nur letzte 100 Einträge behalten
    state.sync_history = state.sync_history.slice(-100);
    saveState(state);
  }

  log("=== Sync Agent beendet ===\n");
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = { 
  loadState, 
  saveState, 
  getAllSkills, 
  initGitRepo, 
  backupSkillDir, 
  syncSkillBidirectional, 
  getHashes 
};

File: perl5/sync_agent.pl
#!/usr/bin/perl
# sync_agent.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/clawhub-git-sync-agent/scripts/sync_agent.py
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use JSON qw(decode_json encode_json);
use File::Path qw(make_path);
use File::Spec;
use File::Copy qw(copy);
use File::Find;
use Digest::MD5 qw(md5_hex);
use POSIX qw(strftime);

# Füge das Skriptverzeichnis zum Suchpfad hinzu
use lib '/home/openclaw/.openclaw/workspace/scripts';
require 'sync_clawhub_git.pl';

my $CLAWHUB_DIR = "/home/openclaw/.openclaw/workspace/skills";
my $GIT_DIR = "/home/openclaw/.openclaw/workspace/git/skills";
my $STATE_FILE = "/home/openclaw/.openclaw/workspace/db/sync_state.json";

# Root directory for backups
my $BACKUP_ROOT = "/home/openclaw/.openclaw/workspace/backups/sync_agent";

sub load_state {
    # Lädt den Sync-State
    if (-f $STATE_FILE) {
        open my $fh, '<', $STATE_FILE or die "Cannot read $STATE_FILE: $!";
        my $json_text = do { local $/; <$fh> };
        close $fh;
        return decode_json($json_text);
    }
    return { sync_history => [], pending => [] };
}

sub save_state {
    # Speichert den Sync-State
    my ($state) = @_;
    my $state_dir = dirname($STATE_FILE);
    make_path($state_dir) unless -d $state_dir;
    open my $fh, '>', $STATE_FILE or die "Cannot write $STATE_FILE: $!";
    print $fh encode_json($state);
    close $fh;
}

sub dirname {
    my ($path) = @_;
    my @parts = split '/', $path;
    pop @parts;
    return join '/', @parts;
}

sub get_all_skills {
    # Findet nur valide Skill-Verzeichnisse in beiden Verzeichnissen.
    my (%clawhub_skills, %git_skills);
    
    opendir(my $dh, $CLAWHUB_DIR) or die "Cannot opendir $CLAWHUB_DIR: $!";
    while (readdir $dh) {
        next if /^\.\.?$/;
        my $dir = "$CLAWHUB_DIR/$_";
        if (-d $dir && !/^[._]/ && -f "$dir/SKILL.md") {
            $clawhub_skills{$_} = 1;
        }
    }
    closedir $dh;
    
    opendir($dh, $GIT_DIR) or die "Cannot opendir $GIT_DIR: $!";
    while (readdir $dh) {
        next if /^\.\.?$/;
        my $dir = "$GIT_DIR/$_";
        if (-d $dir && !/^[._]/ && -f "$dir/SKILL.md") {
            $git_skills{$_} = 1;
        }
    }
    closedir $dh;
    
    my %union = (%clawhub_skills, %git_skills);
    return keys %union;
}

sub init_git_repo {
    # Initialisiert Git-Repo wenn nötig
    my ($skill_path, $skill_name) = @_;
    my $git_dir = "$skill_path/.git";
    if (!-d $git_dir) {
        chdir $skill_path or die "Cannot chdir to $skill_path: $!";
        system("git", "init");
        system("git", "add", ".");
        system("git", "commit", "-m", "Initial commit: $skill_name skill");
        log_msg("Git initialized for $skill_name");
    }
}

sub backup_skill_dir {
    # Creates a timestamped tar.gz backup of a skill directory.
    my ($skill_path, $skill_name) = @_;
    return unless -d $skill_path;
    my $timestamp = strftime "%Y%m%d%H%M%S", localtime;
    my $backup_dir = "$BACKUP_ROOT/$timestamp";
    make_path($backup_dir) unless -d $backup_dir;
    my $archive_name = "${skill_name}_${timestamp}.tar.gz";
    my $archive_path = "$backup_dir/$archive_name";
    system("tar", "-czf", $archive_path, "-C", $skill_path, ".");
    log_msg("Backup created for $skill_name at $archive_path");
}

sub sync_skill_bidirectional {
    # Bidirektionale Synchronisation eines Skills
    my ($skill_name, $dry_run) = @_;
    my $clawhub_path = "$CLAWHUB_DIR/$skill_name";
    my $git_path = "$GIT_DIR/$skill_name";

    # Fall 1: Nur in ClawHub → zu Git
    if (-d $clawhub_path && !-d $git_path) {
        log_msg("NEW in ClawHub: $skill_name → syncing to Git");
        backup_skill_dir($clawhub_path, "${skill_name}_clawhub") unless $dry_run;
        if (sync_to_git($skill_name, $dry_run)) {
            init_git_repo($git_path, $skill_name) unless $dry_run;
            return "synced_to_git";
        }
    }

    # Fall 2: Nur in Git → zu ClawHub
    elsif (-d $git_path && !-d $clawhub_path) {
        log_msg("NEW in Git: $skill_name → syncing to ClawHub");
        backup_skill_dir($git_path, "${skill_name}_git") unless $dry_run;
        if (sync_to_clawhub($skill_name, $dry_run)) {
            return "synced_to_clawhub";
        }
    }

    # Fall 3: In beiden vorhanden → Vergleiche Timestamps
    elsif (-d $clawhub_path && -d $git_path) {
        # --- MODIFIZIERTE LOGIK: Robusterer Datei-Hash-Vergleich ---

        # Stelle sicher, dass beide als gültige Skills validiert werden
        if (!validate_skill($clawhub_path)) {
            log_msg("Validation failed for ClawHub skill: $skill_name", "ERROR");
            return "error";
        }
        if (!validate_skill($git_path)) {
            log_msg("Validation failed for Git skill: $skill_name", "ERROR");
            return "error";
        }

        # Berechne Hashes für clawhub und git
        my $clawhub_hashes = get_hashes($clawhub_path);
        my $git_hashes = get_hashes($git_path);

        if (!hashes_equal($clawhub_hashes, $git_hashes)) {
            log_msg("Content difference detected for: $skill_name");

            # Einfache (aber oft ausreichende) Logik: Wenn clawhub neuer ist, lade hoch.
            # Eine detailliertere Strategie (z.B. welche Version von Git übernehmen)
            # könnte hier implementiert werden, falls nötig.
            # Für jetzt: Wenn sie sich unterscheiden, priorisieren wir ClawHub > Git
            # und aktualisieren Git.

            my $direction = (-M $clawhub_path <= -M $git_path) ? "to-git" : "to-clawhub";
            log_msg("UPDATE: $skill_name → syncing $direction");
            unless ($dry_run) {
                backup_skill_dir($clawhub_path, "${skill_name}_clawhub");
                backup_skill_dir($git_path, "${skill_name}_git");
            }
            my $sync_func = ($direction eq "to-git") ? \&sync_to_git : \&sync_to_clawhub;
            if ($sync_func->($skill_name, $dry_run)) {
                if (!$dry_run && $direction eq "to-git") {
                    chdir $git_path or die "Cannot chdir to $git_path: $!";
                    system("git", "add", ".");
                    my $date_str = strftime "%Y-%m-%d %H:%M", localtime;
                    system("git", "commit", "-m", "Sync from ClawHub content diff: $date_str");
                }
                return ($direction eq "to-git") ? "updated_git" : "updated_clawhub";
            } else {
                log_msg("Failed to sync $skill_name to Git after content diff", "ERROR");
                return "error";
            }
        } else {
            log_msg("Content is identical for: $skill_name");
            return "no_change";
        }
    }

    return "no_change";
}

# --- Hinzufügen dieser Hilfsfunktion ---
sub get_hashes {
    # Erzeugt ein Dictionary von Datei-Hashes für einen Skill-Ordner.
    my ($skill_dir) = @_;
    my %hashes;
    
    find(sub {
        return if -d $_ || /\.git/;
        my $file_path = $File::Find::name;
        open my $fh, '<', $_ or die "Cannot open $_: $!";
        my $content;
        { local $/; $content = <$fh>; }
        close $fh;
        my $relative_path = substr($file_path, length($skill_dir) + 1);
        $hashes{$relative_path} = md5_hex($content);
    }, $skill_dir);
    
    return \%hashes;
}

sub hashes_equal {
    my ($hash1, $hash2) = @_;
    return 0 if keys %$hash1 != keys %$hash2;
    for my $key (keys %$hash1) {
        return 0 if !exists $hash2->{$key} || $hash1->{$key} ne $hash2->{$key};
    }
    return 1;
}

sub main {
    # Hauptfunktion des Sync-Agents
    my $dry_run = grep { $_ eq '--dry-run' } @ARGV;
    log_msg("=== ClawHub ↔ Git Sync Agent gestartet ===");

    my $state = load_state();
    my @all_skills = get_all_skills();
    log_msg("Gefundene Skills: " . scalar(@all_skills));

    my %results = (
        synced_to_git => [],
        synced_to_clawhub => [],
        updated_git => [],
        updated_clawhub => [],
        no_change => [],
        errors => []
    );

    for my $skill (sort @all_skills) {
        eval {
            my $result = sync_skill_bidirectional($skill, $dry_run);
            push @{$results{$result}}, $skill;
        };
        if ($@) {
            log_msg("ERROR syncing $skill: $@", "ERROR");
            push @{$results{errors}}, $skill;
        }
    }

    # Zusammenfassung
    log_msg("\n=== SYNC ZUSAMMENFASSUNG ===");
    log_msg("Neu in Git: " . scalar(@{$results{synced_to_git}}) . " - [" . join(", ", @{$results{synced_to_git}}) . "]");
    log_msg("Neu in ClawHub: " . scalar(@{$results{synced_to_clawhub}}) . " - [" . join(", ", @{$results{synced_to_clawhub}}) . "]");
    log_msg("Git aktualisiert: " . scalar(@{$results{updated_git}}) . " - [" . join(", ", @{$results{updated_git}}) . "]");
    log_msg("ClawHub aktualisiert: " . scalar(@{$results{updated_clawhub}}) . " - [" . join(", ", @{$results{updated_clawhub}}) . "]");
    log_msg("Keine Änderung: " . scalar(@{$results{no_change}}));
    log_msg("Fehler: " . scalar(@{$results{errors}}) . " - [" . join(", ", @{$results{errors}}) . "]");

    # Ein Dry-Run bleibt vollständig nicht-mutierend (abgesehen vom Audit-Log).
    unless ($dry_run) {
        $state->{sync_history} = [] unless exists $state->{sync_history};
        push @{$state->{sync_history}}, {
            timestamp => strftime("%Y-%m-%dT%H:%M:%S", localtime),
            results => \%results
        };
        # Nur letzte 100 Einträge behalten
        splice(@{$state->{sync_history}}, 0, scalar(@{$state->{sync_history}}) - 100) 
            if scalar(@{$state->{sync_history}}) > 100;
        save_state($state);
    }

    log_msg("=== Sync Agent beendet ===\n");
}

main() if !caller;

1;
